### PR TITLE
Added version for REQUIREDSCRIPTS

### PR DESCRIPTION
**Pull request by wooseung-sim**
Added more functionality on 'REQIUREDSCRIPTS' when publishing modules/scripts
It allows user can specify a version of each 'REQIUREDSCRIPTS'. The legacy way that passing only script names are also available.
e.g.> Define 'REQUIREDSCRIPTS' without version (latest version)
.REQUIREDSCRIPTS DummyScript DummyScript2
e.g.>. Define 'REQUIREDSCRIPTS' with specific version
.REQUIREDSCRIPTS DummyScript;2.2.6 DummyScript2;1.0.1

### CODE REVIEW

**Comment by bmanikm** [comment]
@victor-sim 
Thanks for your contribution. please complete the Contribution License Agreement at https://cla.microsoft.com.
From the current code changes, it looks like versioning requirements are validated only during publish operation. What about during install and update operations?
Please note that side-by-side installation is not supported for scripts. How is the versioning requirements going to work with multiple scripts with different dependency script version requirements?

Please consider creating an issue with the design proposal for supporting version ranges for script dependencies.

**Comment by wooseung-sim** [comment]
@bmanikm 
> From the current code changes, it looks like versioning requirements are validated only during publish operation. What about during install and update operations?

Once it passes requiredscript validation, a script will be packaged with 'RequiredVersion' condition. So nuget should handle this, just like module dependencies.

> Please note that side-by-side installation is not supported for scripts. How is the versioning requirements going to work with multiple scripts with different dependency script version requirements?

 I also considered adopting various options of version, such as min/max/required version support by using NuSpec version formatting or using hashtable format like module dependency.
* '(', ')' - exclusive, '[', ']' - inclusive, '\<MinVersion\>,\<MaxVersion\>'

But I just stopped implementing nuspec version format and make it simple to have only required version, as there's no version conflict resolution for dependencies like you said.
As I didn't feel it worth enough without conflict resolution, I intentionally implemented it as is.

If you feel having nuspec version format is worth enough, I can keep going on this and send another pull request

Thank you.

**Comment by wooseung-sim** [comment]
I've created issue #164

**Comment by bmanikm** [comment]
If you would like to just enable the exact version requirement, please consider `<ScriptName>:[<version>]`. With this, later we can enable other version formats similar to the nuspec's dependency version formatting.

Please add required tests as well.

//Cc @JKeithB @KrishnaV-MSFT

**Comment by wooseung-sim** [comment]
@bmanikm 
Thank you for your comment.
I've pushed another commit which includes, supporting nuspec version string format and unit test.
It doesn't support full variations of nuspec version string types as only 4 types of nuspec version strings are supported on PSGet at packaging time. 
>[1.0]  - RequiredVersion
[1.0,2.0] - Min and Max Version
(,1.0] - Max Version
1.0 - Min Version

File: PowerShellGet/PSModule.psm1 [hunk 1]
```diff
@@ -7858,6 +7889,146 @@ function ValidateAndGet-ScriptDependencies
...
 
+function ValidateAndGet-NuspecVersionString
+{
+    param(
+        [Parameter(Mandatory=$true)]
+        [string]
+        $Version
+    )
+
+    $versionPattern = '^((?<MinRule>[\[\(])?((?<MinVersion>[^:\(\[\)\]\,]+))?((?<Comma>[\,])?(?<MaxVersion>[^:\(\[\)\]\,]+)?)?(?<MaxRule>[\]\)])?)$'
```
Comment (bmanikm) [inline PowerShellGet/PSModule.psm1:7900]: >MinVersion [](start = 48, length = 10)

Should we ensure that extracted version string is a proper version? Users can specify [1112], [ewqew323], [1.1.1.1.1], [1.foo], etc.,?

**Review by bmanikm — CHANGES_REQUESTED** [review]
:clock1:

File: PowerShellGet/PSModule.psm1 (hunk 1 above)
Comment (wooseung-sim) [reply PowerShellGet/PSModule.psm1:7900]: ValidateAndGet-NuspecVersionString cmdlet does not validate version string fully. It checks format of the version string.
Such as checking inclusive/exclusive parentheses and/or brackets come together or not, checking existence of version string, detecting invalid or not supported expressions.

And ValidateAndGet-ScriptDependencies cmdlet ensures extracted version string is a proper version by using Find-Script cmdlet, just like it ensure version of dependency modules.

Comment (bmanikm) [reply PowerShellGet/PSModule.psm1:7900]: Okay, thanks!

---
In reply to: [135691066](https://github.com/PowerShell/PowerShellGet/pull/162#discussion_r135691066) [](ancestors = 135691066)

**Review by bmanikm — APPROVED** [review]
:shipit:

**Pull request merged**